<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Prepare data | VCFShiny Description Document</title>
  <meta name="description" content="VCFshiny software Usage Instructions" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Prepare data | VCFShiny Description Document" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="VCFshiny software Usage Instructions" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Prepare data | VCFShiny Description Document" />
  
  <meta name="twitter:description" content="VCFshiny software Usage Instructions" />
  

<meta name="author" content="Tao Chen; Shixue Gou" />


<meta name="date" content="2023-04-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="run-the-app.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">WGS-Shiny Description Document</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#background"><i class="fa fa-check"></i><b>1.1</b> Background</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="index.html"><a href="index.html#the-development-of-sequencing-technology"><i class="fa fa-check"></i><b>1.1.1</b> The development of sequencing technology</a></li>
<li class="chapter" data-level="1.1.2" data-path="index.html"><a href="index.html#whole-genomic-sequencing"><i class="fa fa-check"></i><b>1.1.2</b> Whole Genomic Sequencing</a></li>
<li class="chapter" data-level="1.1.3" data-path="index.html"><a href="index.html#research-significance"><i class="fa fa-check"></i><b>1.1.3</b> Research Significance</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#how-to-start"><i class="fa fa-check"></i><b>1.2</b> How to start</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#requirement"><i class="fa fa-check"></i><b>1.2.1</b> Requirement</a></li>
<li class="chapter" data-level="1.2.2" data-path="index.html"><a href="index.html#how-to-install-shiny-packsge"><i class="fa fa-check"></i><b>1.2.2</b> How to install shiny packsge</a></li>
<li class="chapter" data-level="1.2.3" data-path="index.html"><a href="index.html#how-to-install-and-run-vcfshiny"><i class="fa fa-check"></i><b>1.2.3</b> How to install and run VCFshiny</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="prepare-data.html"><a href="prepare-data.html"><i class="fa fa-check"></i><b>2</b> Prepare data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="prepare-data.html"><a href="prepare-data.html#source-of-vcf-input-data"><i class="fa fa-check"></i><b>2.1</b> Source of VCF input data</a></li>
<li class="chapter" data-level="2.2" data-path="prepare-data.html"><a href="prepare-data.html#source-of-txt-input-data"><i class="fa fa-check"></i><b>2.2</b> Source of TXT input data</a></li>
<li class="chapter" data-level="2.3" data-path="prepare-data.html"><a href="prepare-data.html#input-data-name-requirements"><i class="fa fa-check"></i><b>2.3</b> Input Data name requirements</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="prepare-data.html"><a href="prepare-data.html#input-file-name-requirements"><i class="fa fa-check"></i><b>2.3.1</b> Input file name requirements</a></li>
<li class="chapter" data-level="2.3.2" data-path="prepare-data.html"><a href="prepare-data.html#input-compress-files-requirements"><i class="fa fa-check"></i><b>2.3.2</b> Input compress files requirements</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="prepare-data.html"><a href="prepare-data.html#example-data-set-in-vcfshiny-software"><i class="fa fa-check"></i><b>2.4</b> Example data set in VCFshiny software</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="run-the-app.html"><a href="run-the-app.html"><i class="fa fa-check"></i><b>3</b> Run The App</a>
<ul>
<li class="chapter" data-level="3.1" data-path="run-the-app.html"><a href="run-the-app.html#part-1-analysis-part"><i class="fa fa-check"></i><b>3.1</b> Part 1: Analysis Part</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="run-the-app.html"><a href="run-the-app.html#module-1-main-page"><i class="fa fa-check"></i><b>3.1.1</b> Module 1: Main Page</a></li>
<li class="chapter" data-level="3.1.2" data-path="run-the-app.html"><a href="run-the-app.html#module-2-data-input"><i class="fa fa-check"></i><b>3.1.2</b> Module 2: Data Input</a></li>
<li class="chapter" data-level="3.1.3" data-path="run-the-app.html"><a href="run-the-app.html#module-3-venn-diagram"><i class="fa fa-check"></i><b>3.1.3</b> Module 3: Venn Diagram</a></li>
<li class="chapter" data-level="3.1.4" data-path="run-the-app.html"><a href="run-the-app.html#module-4-variants-number"><i class="fa fa-check"></i><b>3.1.4</b> Module 4: Variants Number</a></li>
<li class="chapter" data-level="3.1.5" data-path="run-the-app.html"><a href="run-the-app.html#module-5-snv-analysis"><i class="fa fa-check"></i><b>3.1.5</b> Module 5: SNV Analysis</a></li>
<li class="chapter" data-level="3.1.6" data-path="run-the-app.html"><a href="run-the-app.html#module-6-indel-analysis"><i class="fa fa-check"></i><b>3.1.6</b> Module 6: Indel Analysis</a></li>
<li class="chapter" data-level="3.1.7" data-path="run-the-app.html"><a href="run-the-app.html#module-7-circos-plot"><i class="fa fa-check"></i><b>3.1.7</b> Module 7: Circos Plot</a></li>
<li class="chapter" data-level="3.1.8" data-path="run-the-app.html"><a href="run-the-app.html#module-8-genomic-feature"><i class="fa fa-check"></i><b>3.1.8</b> Module 8: Genomic Feature</a></li>
<li class="chapter" data-level="3.1.9" data-path="run-the-app.html"><a href="run-the-app.html#module-9-highest-variant-genes"><i class="fa fa-check"></i><b>3.1.9</b> Module 9: Highest Variant Genes</a></li>
<li class="chapter" data-level="3.1.10" data-path="run-the-app.html"><a href="run-the-app.html#module-10-cancer-driver-genes"><i class="fa fa-check"></i><b>3.1.10</b> Module 10: Cancer Driver Genes</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="run-the-app.html"><a href="run-the-app.html#part-2-annotation-part"><i class="fa fa-check"></i><b>3.2</b> Part 2: Annotation Part</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="run-the-app.html"><a href="run-the-app.html#module-1-annovar-annotation"><i class="fa fa-check"></i><b>3.2.1</b> Module 1: Annovar Annotation</a></li>
<li class="chapter" data-level="3.2.2" data-path="run-the-app.html"><a href="run-the-app.html#module-2-variantannotation-annotation"><i class="fa fa-check"></i><b>3.2.2</b> Module 2: VariantAnnotation Annotation</a></li>
<li class="chapter" data-level="3.2.3" data-path="run-the-app.html"><a href="run-the-app.html#module-3-download-annovar-database"><i class="fa fa-check"></i><b>3.2.3</b> Module 3: Download Annovar Database</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="conclusion.html"><a href="conclusion.html"><i class="fa fa-check"></i><b>4</b> Conclusion</a></li>
<li class="chapter" data-level="5" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>5</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">VCFShiny Description Document</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="prepare-data" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Prepare data<a href="prepare-data.html#prepare-data" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this chapter, we describe how to prepare two different input datasets, where the annotation part of VCFshiny can only input unannotated VCF files, and the analysis part of VCFshiny can read annotated VCF and TXT files.</p>
<div id="source-of-vcf-input-data" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Source of VCF input data<a href="prepare-data.html#source-of-vcf-input-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The Variant Call Format (VCF) is used to record gene sequence variations. It is also the first file format to be understood for genome population correlation analysis. The file is divided into two main parts: the Header comment section, which begins with <em>#</em>, and the body section.</p>
</div>
<div id="source-of-txt-input-data" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Source of TXT input data<a href="prepare-data.html#source-of-txt-input-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>TXT file is one of the standard storage formats after mutation data annotation. Because some users cannot annotate the steps, We provide Annovar (<a href="references.html#ref10">Wang K, Li M, Hakonarson H. 2010</a>) and VariantAnnotation (<a href="reference.html#11">Obenchain, V., et al. 2014</a>) two methods of annotation in the VCFshiny, Users can choose to annotate themselves or use the tool.</p>
</div>
<div id="input-data-name-requirements" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Input Data name requirements<a href="prepare-data.html#input-data-name-requirements" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="input-file-name-requirements" class="section level3 hasAnchor" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> Input file name requirements<a href="prepare-data.html#input-file-name-requirements" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-7"></span>
<img src="figure/Input_File_Format.png" alt="File-Format" width="80%" />
<p class="caption">
Figure 2.1: File-Format
</p>
</div>
<ol style="list-style-type: decimal">
<li>The first box represents the sample name, which can be the group of experiments and the number of repetitions, connected by the character <em>“-”</em> or *“_“*.</li>
<li>The second box represents the data type, which can be snp or indel data. When snp and indel are not classified in the data, <strong>this box is not necessary</strong>.</li>
<li>The third box represents the data format, which can be vcf files, vcf. gz compressed files, and <em>Annovar</em> annotated TXT files.</li>
<li>The contents of the three boxes are connected by <em>“.”</em>.</li>
</ol>
</div>
<div id="input-compress-files-requirements" class="section level3 hasAnchor" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> Input compress files requirements<a href="prepare-data.html#input-compress-files-requirements" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Before uploading the data to VCFshiny, that needs to be compressed. The following are the naming requirements for the compressed folder.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-8"></span>
<img src="figure/Fold_Format.png" alt="Fold_Format" width="80%" />
<p class="caption">
Figure 2.2: Fold_Format
</p>
</div>
<ol style="list-style-type: decimal">
<li>The compressed file name must be the same as the name of the compressed folder.</li>
<li>The compressed file can be in <code>*.tar.gz</code> or <code>*.zip</code> format.</li>
</ol>
</div>
</div>
<div id="example-data-set-in-vcfshiny-software" class="section level2 hasAnchor" number="2.4">
<h2><span class="header-section-number">2.4</span> Example data set in VCFshiny software<a href="prepare-data.html#example-data-set-in-vcfshiny-software" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We provide a built-in dataset that can be used to explore VCFshiny. The data set can be directly loaded into the APP by clicking the button <em>“Use example data?”</em> of the data input module (Figure S4-II). The built-in dataset was derived from the sequencing data of published articles, including a control group and three experimental groups, with three replicates in each group (<a href="references.html#ref12">Liang Y, Xie J, Zhang Q, et al. 2022</a>). The sequencing data was first mapping to the reference, followed by gatk Variants calling and Annovar annotations.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="run-the-app.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/01-prepare_data.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
